src/table/system.rs

```diff
@@ -1,68 +1,228 @@
     }
 
     /// Returns the standard error protocol.
     pub fn stderr(&self) -> &mut text::Output {
-        unsafe { &mut *self.stderr }
+        let stderr_ptr = self.table.stderr as *const _ as *mut _;
+        unsafe { &mut *stderr_ptr }
     }
 
-    /// Returns the config table entries, a linear array of structures
-    /// pointing to other system-specific tables.
-    pub fn config_table(&self) -> &[cfg::ConfigTableEntry] {
-        unsafe { slice::from_raw_parts(self.cfg_table, self.nr_cfg) }
+    /// Access runtime services
+    pub fn runtime_services(&self) -> &RuntimeServices {
+        self.table.runtime
     }
+
+    /// Access boot services
+    pub fn boot_services(&self) -> &BootServices {
+        unsafe { &*self.table.boot }
+    }
+
+    /// Exit the UEFI boot services
+    ///
+    /// After this function completes, the UEFI boot services are shut down and
+    /// cannot be used anymore. Only run-time services can be used. We model
+    /// this by consuming the SystemTable<Boot> view and returning a more
+    /// restricted SystemTable<Runtime> view as an output.
+    ///
+    /// The handle passed must be the one of the currently executing image,
+    /// which is received by the entry point of the UEFI application. In
+    /// addition, the application **must** retrieve the current memory map, and
+    /// pass the associated MemoryMapKey so that the firmware can check that it
+    /// is up to date.
+    ///
+    /// If the application's memory map is not up to date, then the firmware
+    /// ends up in an awkward situation where it may have already shut down some
+    /// of the boot services, but it wants to give the application a chance to
+    /// fetch a newer memory map. This is modeled in the API by the
+    /// `retry_handler`, a user-provided functor which is given as input...
+    ///
+    /// - The new size of the memory map
+    /// - Access to the `memory_map` boot service (as described in the
+    ///   `BootServices` table)
+    ///
+    /// The functor _may_ attempt to use this information to fetch the memory
+    /// map again and return the associated key. It will be repeatedly called
+    /// by `exit_boot_services` implementation as long as it does so but the
+    /// output MemoryMapKey is stale. If the functor chooses to return `None`
+    /// instead, `exit_boot_services` will terminate with an error.
+    ///
+    /// If `exit_boot_services` succeeds, it will return a runtime view of the
+    /// system table which more accurately reflects the state of the UEFI
+    /// firmware following exit from boot services.
+    pub fn exit_boot_services(
+        self,
+        image: Handle,
+        mmap_key: MemoryMapKey,
+        mut retry_handler: impl FnMut(
+            // Size of the new memory map
+            usize,
+            // Access to the `memory_map` boot service
+            &mut FnMut(&mut [u8]) -> Result<(MemoryMapKey, MemoryMapIter)>,
+        ) -> Option<MemoryMapKey>,
```

Comment:
Come to think of it, I'm wondering if the full generality of this retry_handler is needed, or if making the high-level exit_boot_services simply take an &mut [u8] as input and return the memory map as an additional output wouldn't be the right call.

After all, I'm not sure what an UEFI application would want to do in this handler besides querying the memory map again (if there's enough space) or giving up (if the new memory map is too large). Global services like logging or memory allocation would be off-limit due to the partial boot services shutdown, and invalidating the user's MemoryMapIter without returning the new one is not a nice interface design.

Reply:
It does seem logical to only support the correct behavior. I've checked both the Linux kernel and `syslinux` and they do not attempt to do anything funny in such a loop.